### PR TITLE
converted sorter.lic from GS to DR

### PR DESCRIPTION
**Pull request by Hazado**
addresses this issue - https://github.com/rpherbig/dr-scripts/issues/163

### CODE REVIEW

**Comment by Sarvatt** [comment]
guess i would need to run this in game to make sense of it because it doesnt from reading the code. we dont use gameobj-data.xml in dr and i imagine its completely broken on our side

**Comment by Sarvatt** [comment]
the equivalents to that is stored in like data/base-items.yaml in dr, and the other base files

**Comment by Hazado** [comment]
Doesnt do the various categories that GS does.  Was away that we had something the equivalent of gameobj-data.xml

Got the basic sorting and one category put in. So some more work needed on it.

**Comment by Hazado** [comment]
Basic sorting of categories is in based off of the base-items.yaml

Im open to better implementations.

**Comment by Hazado** [comment]
Improved the list_to_array(list) and get_noun(long_name) regex.
Helps with capturing items correctly.

get_noun regex before would tell me this noun was marked and not runestone - a sunstone runestone marked with a symbol for Compost

list_to_array would sort items with "and" in their name as two items.
example: an ivory and jade plaid kilt fastened with a silver ivy wreath pin

**Comment by jandersson** [comment]
You have some line ending issues it looks like. The diff shows the entire file as changed.

**Comment by Hazado** [comment]
yeah, let me try again

**Comment by Hazado** [comment]
Yeah, dont know what that is about. I got a fresh copy of common.lic and made my changes. Still does it.

**Comment by jandersson** [comment]
One thing to do is set `git config --global core.autocrlf false` if you use git cli

**Comment by jayRyan24** [comment]
This is awesome. I will use it tonight and test it, but thanks a lot for adding this.

File: sorter.lic [hunk 1]
```diff
@@ -0,0 +1,285 @@
+=begin
```
Comment (rpherbig) [inline sorter.lic]: 💡 Consider documenting the script on elanthipedia:
```
=begin
  Documentation: https://elanthipedia.play.net/Lich_script_development#sorter
=end
```

File: sorter.lic [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+        0.3 (2014-11-21):
+            testing fix for squelching containers that can't be sorted
+        0.2 (2014-11-10):
+            don't display container contents twice on Lich 4.6
+
+=end
+
+custom_require.call(%w[common])
+
+if script.vars[1] =~ /^width=([0-9]+|nil)$/
```
Comment (rpherbig) [inline sorter.lic]: 💡 Our args system would make this easier to use

Comment (Hazado) [reply sorter.lic]: That option is to do columned sorting and it doesn't usually look right when you do it. I'm not a fan of the look but I left it in because all the work to get it running is already done. Might as well convert it to the arg system.

**Comment by rpherbig** [comment]
To be clear, @Hazado , anything I tag with 💡 is just an idea, it does not require action.

If I see something dire/required, I'll use ❌

File: sorter.lic (hunk 2 above)
Comment (Hazado) [reply sorter.lic]: Can add an arg for ignoring sorting groups in base-items.yaml. Currently have it hard-coded to ignore lootables and trash.

**Comment by rpherbig** [comment]
Thank you for fixing the line endings!

File: sorter.lic (hunk 1 above)
Comment (Hazado) [reply sorter.lic]: Done

File: sorter.lic (hunk 2 above)
Comment (Hazado) [reply sorter.lic]: args added

**Comment by Hazado** [comment]
Think its ready for general use now.
base-item.yaml will take a while to add various categories though.

added some args - details at https://elanthipedia.play.net/Lich_script_development#sorter

```
sorter:
  width: 0
  mute_old_inventory: true
  sort_inv_command: true
  sort_look_items_command: true
  ignore_categories: lootables|trash
```

File: profiles/base.yaml [hunk 3]
```diff
@@ -298,8 +298,8 @@ cambrinth_cap: 32
 stored_cambrinth: false
 # used with use_auto_mana. Multiplies your discern limit by this number to get a safe mana amount.
 prep_scaling_factor: 0.85
-# used with use_auto_mana. How often it will check discerns. Lower circle/ranks should check more often.
-check_discern_timer: 24
+# used with use_auto_mana. How often it will check discerns in hours. Lower circle/ranks should check more often.
```
Comment (jayRyan24) [inline profiles/base.yaml]: This is a merge issue. This and like 302.

**Comment by jayRyan24** [comment]
```
  a silver-bound dark leather ankle sheath
     -a tyrium skinning knife with a mistglass handle
     -a folded steel bushcraft knife with a blackwater jet handle
  some weathered leather sandals with burnished brass buckles
 
Roundtime:  10 secs
 
[Type INVENTORY HELP for more options]
>inv
[Type INVENTORY HELP for more options]
>
--- Lich: error: comparison of Array with Array failed
	sorter:109:in `sort'
	sorter:109:in `block in _script'
--- Lich: sorter has exited.
```

**Comment by jayRyan24** [comment]
So, I ran sorter and did `inv list` and `inv`, inv breaks it.

**Comment by jayRyan24** [comment]
look in my pack is working fine
```
;sort
--- Lich: sorter active.
>look in my pack
In the hunting pack:
other (48): 
    rugged backpack
    bent silver badge
    brushed gold badge
    cracked gold badge
    gleaming steel badge
    canvas bag
    large hex bag
    master blacksmithing book
    apprentice tinkering book
    chamois cloth
    cleaning cloth
    copper locust earcuff
    crocodile tooth earring
    simple paper fan
    lightning amethyst flask
    borax flux
    glass globe
    oak hilt
    leather warrior's kit
    polished knitting needles
    tiny tin nugget (2)
    flask of oil
    festive memory orb
    tiny copper memory orb (2)
    large deed packet
    lampblack paintstick
    sheet of cobalt paper
    corn husk poppet
    suede gem pouch
    tattered rag
    twisted cambrinth rod
    bone saw
    grey ice skates
    small snowman
    lot of other stuff
    chakrel toe-ring
    stuffed mistletoad tote
    kertig warhorn
    blue weevil
    worn (7)

box (2): 
    misshaped deobar trunk
    sturdy brass trunk

gem (17): 
    kyanite stones (5)
    jadeite stones (12)

scroll (6): 
    piece of hhr'lav'geluhh bark
    smudged parchment
    papyrus roll
    tattered scroll
    yellowed scroll
    clay tablet

valuable (10): 
    tattered map (7)
    blue notebook
    brown paper-wrapped package (2)

armor (2): 
    quilted doublet
    half mask

ammunition (1): 
    small belzune rock
```

**Comment by jayRyan24** [comment]
25 width worked for me:
```# Sorter
sorter:
  width: 25
  mute_old_inventory: true
  sort_inv_command: true
  sort_look_items_command: true
  ignore_categories: lootables|trash```

File: data/base-items.yaml [hunk 4]
```diff
@@ -274,3 +274,177 @@ lumber_types:
 - tamarak # alterations
 - tamboti # ???
 - yew # longbows
+
```
Comment (jayRyan24) [inline data/base-items.yaml]: What do you guys think about a base-sorter file?

Comment (jayRyan24) [inline data/base-items.yaml]: We need a clothing section as well!

**Comment by jayRyan24** [comment]
This is a brilliant addition @Hazado!

**Comment by Hazado** [comment]
> 
> 
> ```
>   a silver-bound dark leather ankle sheath
>      -a tyrium skinning knife with a mistglass handle
>      -a folded steel bushcraft knife with a blackwater jet handle
>   some weathered leather sandals with burnished brass buckles
>  
> Roundtime:  10 secs
>  
> [Type INVENTORY HELP for more options]
> >inv
> [Type INVENTORY HELP for more options]
> >
> --- Lich: error: comparison of Array with Array failed
> 	sorter:109:in `sort'
> 	sorter:109:in `block in _script'
> --- Lich: sorter has exited.
> ```
did you use the latest commit i did, might fix that

**Comment by jayRyan24** [comment]
I have the latest.

**Comment by Hazado** [comment]
might be because the common.lic changes get rewritten when you run dependency, give me a sec to upload a version that doesnt need them and try that

**Comment by jayRyan24** [comment]
Yes, that's it. I ran `;common-item` but not `;common`. Thanks!

**Comment by jayRyan24** [comment]
```
Inventory:
other (31): 
    chakrel gear attached to a curving metal band
    dusty highwayman's thigh bag with thin leather bindings
    tiny glass bottle
    rugged forger's belt wrapped with fraying grey bourde
    cobalt hide haversack scored with brands
    weathered leather sandals with burnished brass buckles
    tight bundle
    multi-strapped carryall
    dull serpent earcuff (2)
    grimy steel ring bearing the crest of the Barbarians' guild
    platinum jadeite gwethdesuan
    damaska boar-hide chest harness
    large hunting pack crafted from wyvern hide
    ornate platinum brooch set with an orichalcum icosahedron
    engraved war hammer hanging from it
    kadepa ring helm crafted from tempered links
    shimmering outfit
    suede gem pouch
    mammoth-bone thigh quiver
    ornate haralun ring
    lockpick ring
    kadepa scale gloves crafted from tempered scales
    coiled black taipan covered in imbricated agonite scales
    heavily-girdled warrior's loincloth intricately tooled with scorpions
    silver-bound dark leather ankle sheath
    chakrel toe-ring wrought like a tiny spider
    intricate tyrium hand suspended from a cord of darkly glistening spidersilk
    diamond-hide wristcuff pierced with crystal-bone spikes
    puffy-white targe sealed with protective wax
    pair of copper zills

armor (2): 
    padded titanese hauberk
    light kadepa plate mask crafted from tempered plates

gem (2): 
    blackened iron rings beaded with golden amber
    sanowret crystal

weapon (2): 
    functional sword belt with a haralun sterak axe
    platinum wrist knife with a ruby-capped pommel

brawling (1): 
    polished steel parry stick
```

**Comment by Hazado** [comment]
really wish simu did xml for items...going to be a huge regex the way its going right now

**Comment by Hazado** [comment]
think it might be easier to make some categories overwrite others, like weapons, clothing , armor has priority over metal and lumber.

**Comment by Hazado** [comment]
lol or do an item search on elanthipedia and include every item for each category

only 1719 light edged weapons

File: data/base-items.yaml (hunk 4 above)
Comment (Hazado) [reply data/base-items.yaml]: probably better way to go about it in the end

Comment (rpherbig) [reply data/base-items.yaml]: Yes

**Comment by jayRyan24** [comment]
I think you shouldn't strive for every item to be accounted for, but rather just the most frequent adjectives and nouns.

**Comment by jayRyan24** [comment]
Also, if it's a sword belt, go with the last known noun. So, the category for this would be clothes and not weapon, type of thing.

**Comment by jayRyan24** [comment]
Also, if all this thing did was list everything between the commas, it would be brilliant as well. The jumbled shit is the problem, at least for my old ass. 👯‍♂

**Comment by Hazado** [comment]
Set your width to zero

**Comment by jayRyan24** [comment]
Oh, I meant in general if this sorter only listed everything it would be worth it, the jumbled text that is just the default `look in my pack` is what the real problem is for me. 25 worked well.

**Comment by Hazado** [comment]
sorting alphabetically is no problem, finding the noun of the item is a bit harder because of the added flavour text. in GS they give every item as an xml with the noun, so its a lot easier to parse. here we need to run it through regex to do that. hence the regex is going to get huge to do that. going to add a debug so it easier to figure out why something went to weapons or gems etc

**Comment by jayRyan24** [comment]
I see, well, that's a bummer.

**Comment by Hazado** [comment]
debug added, ;vars set sorter_debug=true

**Comment by Hazado** [comment]
went through all the weapons and made a regex to grab the nouns of them.
First go through... only 3514 weapons total

Tommorow is the armor, then ill work on a base-sorter.yaml

**Comment by Hazado** [comment]
I think its ready to be added now.
The regex is a huge mess, but it works.
Will probably need to separate out each section of the regex and then combine them with a smaller regex.

basically its (prefix 1)(prefix 2)(connecting)(suffix)(special cases)

File: profiles/base.yaml (hunk 3 above)
Comment (rpherbig) [reply profiles/base.yaml]: @Hazado can you merge master into your branch to make sure this is resolved?

**Comment by Hazado** [comment]
Think it's fixed, not so good at the github yet

**Comment by rpherbig** [comment]
@jandersson @jayRyan24 have you reviewed the latest?

**Comment by Hazado** [comment]
It works relatively well at this point, the regex is always going to be an issue until DR devs implement item links like GS has, if ever.
It captures a majority of the games items correctly but will probably be messing with the regex till the end of time.

File: sorter.lic [hunk 5]
```diff
@@ -0,0 +1,179 @@
+=begin
+  Documentation: https://elanthipedia.play.net/Lich_script_development#sorter
+=end
+
+custom_require.call(%w[common])
+
+@settings = get_settings
+@settings.sorter['ignore_categories'] ||= "nil"
```
Comment (jandersson) [inline sorter.lic]: Why the string nil and not actual nil?

Also since lich overrides nil, this isn't necessary - if you try to lookup a key that doesn't exist, you get `nil` back
```
>;e a_hash = {}; echo a_hash['ignore_categories']        
--- Lich: exec1 active.                                  
[exec1: ]                                                
```

**Comment by jandersson** [comment]
Since its a new script, I think its alright getting it out there as long as it runs. Itll be easier to get feedback as well. Looks really interesting!

File: data/base-sorting.yaml [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+- diamond$
+- crystal$
+- emerald$
+- ruby$
+- tourmaline$
+- tanzanite$
+- jade$
+- ivory$
+- sunstone$
+- iolite$
```
[the remark this answers is not in the document]
Comment (Hazado) [reply data/base-sorting.yaml]: its so that if the item contains the word "ivory" it doesn't show the item as a gem, only items that end in "ivory"

File: sorter.lic (hunk 5 above)
Comment (Hazado) [reply sorter.lic]: Gave me an error if ignore_categories was blank. Using just nil also gave an error, a string "nil" was easier fix.

File: common.lic [hunk 7]
```diff
@@ -220,7 +220,13 @@ module DRC
...
...s|mirror-finished|mottled|multiple woods|naphtha|oak|oblong sanguine|one|onyx buttons|opposing images|overlapping|pale cerulean|pallid links|pastel-hued|pins|pitted (?:black iron|steel)|plush velvet|polished (?:bronze|hemlock|steel)|raccoon tails|ram's horns|rat pelts|raw|red and blue|rich (?:purple|golden)|riveted bindings|roughened|rowan|sanguine thornweave|scattered star|scorch marks|sculpted|shadows|shark cartilage|shifting (?:celadon|shades)|shipboard|(?:braided |cobalt |deep black |desert-tan |dusky red Taisidon |ebony |exquisite spider|fine leaf-green |flowing night|glimmering ebony |heavy |marigold |pale gold marquisette and virid |rich copper |spiral-braided |steel|unadorned black Musparan )?silk(?:cress)?|(?:coiled |shimmering )?silver(?:steel| and |y)?|sirese blue spun glitter|six crossed|slender|small bones|smoothly interlocking|snow leopard|soft brushed|somber black|sprawled|sun-bleached|steel links|stones|strips of|sunny yellow|teardrop plates|telothian|the|tiny (?:golden|indurium|scales|skull)|tightly braided|tomiek|torn|twists|two|vibrant multicolored|viscous|waves of|weighted|well-cured|white ironwood|windstorm gossamer|wintry faeweave|woven diamondwood))\b.*/, '')
```
[the remark this answers is not in the document]
Comment (Hazado) [reply common.lic]: the base-sorting.yaml maybe?

[the remark this answers is not in the document]
Comment (jandersson) [reply common.lic]: That sounds like a good spot to me

File: data/base-sorting.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jandersson) [reply data/base-sorting.yaml]: Is there any way to programaticaly append the `$` and read from gem_list in base-items? It just seems redundant. I dont suggest this as a blocker, only refactoring at some point. because its the same data

File: sorter.lic (hunk 5 above)
Comment (jandersson) [reply sorter.lic]: Ok, are you the author of the sorter script?

File: common.lic (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hazado) [reply common.lic]: Not exactly sure how to add it to a yaml, get an error parsing the yaml when I try

[the remark this answers is not in the document]
Comment (jandersson) [reply common.lic]: `get_data('items').gem_nouns`

if you want to play with it ingame try out `;e echo get_data('items').gem_nouns`

[the remark this answers is not in the document]
Comment (Hazado) [reply common.lic]: What I can't figure out is how to add the regex to a yaml. It doesn't parse correctly when I try.

[the remark this answers is not in the document]
Comment (jandersson) [reply common.lic]: You've added regex yaml files you submitted

[the remark this answers is not in the document]
Comment (jandersson) [reply common.lic]: What im wondering is when you read the settings in the sorter script, can you append the the end of line markers?

File: data/base-sorting.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jayRyan24) [reply data/base-sorting.yaml]: Shouldn't block, but it's something to consider. I think even adding a warning comment to this section to "UPDATE HERE AND SORTER WHEN MAKING CHANGES" would be beneficial.

**Comment by jayRyan24** [comment]
Working so well for me. Very great addition @Hazado

**Comment by jayRyan24** [comment]
```
[hunting-buddy]>rummage /B my haversack
You rummage through a large hunting pack crafted from wyvern hide looking for boxes:
box (2): 
    sturdy pine skippet
    crooked steel strongbox
>
| hunting-buddy: ***STATUS*** Skipping hunt because of current boxes is more than the minimum.
Feltzer visibly churns with an inner rage!
>
[hunting-buddy]>inv combat
--- Lich: error: undefined method `join' for "Crossing":String
	sorter:87:in `block (2 levels) in _script'
	sorter:83:in `each'
--- Lich: sorter has exited.
```

**Comment by Hazado** [comment]
> 
> 
> ```
> [hunting-buddy]>rummage /B my haversack
> You rummage through a large hunting pack crafted from wyvern hide looking for boxes:
> box (2): 
>     sturdy pine skippet
>     crooked steel strongbox
> >
> | hunting-buddy: ***STATUS*** Skipping hunt because of current boxes is more than the minimum.
> >
> [hunting-buddy]>inv combat
> --- Lich: error: undefined method `join' for "Crossing":String
> 	sorter:87:in `block (2 levels) in _script'
> 	sorter:83:in `each'
> --- Lich: sorter has exited.
> ```

That is a very weird one, not sure where Crossing came into the string at all.

File: data/base-sorting.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hazado) [reply data/base-sorting.yaml]: I got it to append $ to the end of everything now without it being in the string.

Some of the regex stuff needs to stay in.
Cambrinth has .*
some of the boxes and other things need \b around them

Should make it easier to update now.

File: common.lic (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hazado) [reply common.lic]: I appended the end of line marks at the join on line 87.
Uploaded the new code.

File: data/base-sorting.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hazado) [reply data/base-sorting.yaml]: Could also change line 79 to add gem_nouns from items instead of having it in both places.
test script for proof of concept.

;e test1 = get_data('sorting').to_h; test2 = get_data('items').to_h; echo test1.merge(test2.keep_if{|key, value| key =~ /gem_nouns|valuables/})

File: data/base-sorting.yaml [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+- white vellum
+- hhr'lav'geluhh bark
+- smudged parchment
+- tattered papyrus
+- ostracon
+- seishaka leaf
+- papyrus roll
+- moldy papyrus
+
+gem_nouns:
```
[the remark this answers is not in the document]
Comment (Hazado) [reply data/base-sorting.yaml]: Fixed with newest version.
Grabs gem_nouns and scroll_nouns from base-items.yaml
line 80 has how

File: data/base-sorting.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hazado) [reply data/base-sorting.yaml]: Added to the latest version

**Comment by jayRyan24** [comment]
I think your build is failing due to issues with Binu's changes.

**Comment by Hazado** [comment]
> I think your build is failing due to issues with Binu's changes.

Yes, think every build is going to fail until whatever is wrong is fixed.

**Comment by BinuDR** [comment]
Yep.
```  1) Failure:
TestDRCA#test_ritual_with_skip_retreat [/home/travis/build/rpherbig/dr-scripts/test/test_common-arcana.rb:77]:
Expected script to send messages.
Expected ["stance set 100 0 80", "prepare SPELL 1", "remove my staff", "invoke my staff", "wear my staff", "cast"] to be empty.
```

**Comment by BinuDR** [comment]
It's because the check for method is looking for the old hard coded value.  I made it use a yaml setting instead.  The check needs updating

**Comment by BinuDR** [comment]
This wouldn't be an issue if you updated your fork and resubmitted the changes though, it's happening because it's including all of those other pull requests.

**Comment by Hazado** [comment]
Resubmitting
